### ISSUE TITLE
Cannot connect with peripheral

### ISSUE THREAD

**Issue by sewerynplazuk** [nordicsemi/IOS-DFU-Library#35]
Hi,
I am developing application and trying to use this library to implement FOTA into it. I am using Swift 3.0 and testing everything on device with iOS 10. Peripheral is based on nRF51822 SoC. 

Unfortunatelly library cannot connect with peripheral - stucks on "Connecting".

I tried running initiator.start() method in different conditions (connected with the device, disconnected with the device, in DFU mode, in normal mode) nothing works. Do you have any idea what am I missing?

bluetooth.manager and bluetooth.connectedPeripheral are not nil.

My implementation:

```
        let bluetooth = SensorManager.sharedInstance
        self.downloadUpdateFile(uuid: forUUID) { (localFirmwareUrl) in
            if let selectedFirmware = DFUFirmware(urlToZipFile: localFirmwareUrl) {
                let initiator = DFUServiceInitiator(centralManager: bluetooth.manager, target: bluetooth.connectedPeripheral).withFirmwareFile(selectedFirmware)

                initiator.packetReceiptNotificationParameter = 12
                initiator.forceDfu = true
                initiator.logger = self
                initiator.delegate = self
                initiator.progressDelegate = self
                print("START")
                let controller = initiator.start()
            }
        }
```

**Comment by mostafaberg** [comment]
This seems to be related to a Bug in iOS10, it has been resolved from Beta 2 and above, can you verify after installing the beta OS ?

**Comment by sewerynplazuk** [comment]
Do you mean iOS 10.0 build 2 or 10.1 build 2?

**Comment by mostafaberg** [comment]
That's `10.1`, sorry didn't mention the exact version :)

**Comment by sewerynplazuk** [comment]
I will verify it as soon as I can and let you know.

**Comment by sewerynplazuk** [comment]
I just tested app on device with iOS10.1 beta 4 and it still stucks at "Connecting".

**Comment by mostafaberg** [comment]
Sorry that you went through the update with no success, can you send the logger output so I can investigate the issue?

**Comment by sewerynplazuk** [comment]
I would but there is no logger output. "Connecting" is the only print I got while running the application.

**Comment by marcussfu** [comment]
we have the same issue too, but we skip centralManager.connect(didConnect and didDiscoverServices), then call self.didDiscoverDFUService(false) in DFUExecutor.start().

As a result, we can update the firmware successfully.

However,we don't know whether the method we mentioned above will cause another issue.
@mostafaberg @sewerynplazuk

**Comment by mostafaberg** [comment]
@sewerynplazuk  There will always be log events, if you don't see them then you probably haven't setup the logger delegate properly, have you implemented the logger delegate protocol on that class ?

Another thing to try is to disconnect the peripheral before doing that, the library assumes the peripheral is disconnected, it might end up getting stuck on connecting, try to disconnect first and see, we have a fix for that on the way that'll be in the next release.

@marcussfu it is a workaround that involves modifying the source code of the library, which makes it hard for us to support, I would try to do the same thing too, disconnect first and then initiate the DFU process and see if it works out

**Comment by sewerynplazuk** [comment]
@marcussfu I'll try it, thanks.
@mostafaberg I will check logger once more and give you feedback but you have to wait until Monday.

I tried running start() method while peripheral was disconnected - it's not the case.

**Comment by marcussfu** [comment]
@sewerynplazuk 
Forgot to mention, we first connect bluetooth with the device before DFUExecutor.start(),  then skip centralManager.connect, and call self.didDiscoverDFUService(false).

**Comment by mostafaberg** [comment]
Ok keep me updated !

**Comment by sewerynplazuk** [comment]
@mostafaberg, I double checked everything and logger delegate is properly set although it just don't print any logs. Method logWith() hasn't fired even once during my test. 

Workaround provided by @marcussfu works great - thanks once again. I am able to perform update now (logs are working too).

**Comment by mostafaberg** [comment]
A bit strange to hear about that logging issue, will look into that soon, 
in the next version as i said i'll include the fix as @marcussfu says by making sure things work when the peripheral is already connected.

thanks for your patience :)

**Comment by DannyWithJoy** [comment]
@mostafaberg , I faced the same issue today. It took me 2 hours to find the reason.
In my OC code, this is not OK:
`[self.initiator start];`
this is OK:
`self.controller = [self.initiator start];`
I'm guess that:
`DFUServiceController` object return by `-[DFUServiceInitiator start];` has been autoreleased later, so after `conecting...`, `DFUServiceController` object no looger exsit.

It's my case. I hope it will give you some help.

**Comment by mostafaberg** [comment]
@DannyWithJoy yes this is the expected behaviour since ObjC Uses ARC so you need to hold a reference to the controller object.

This is a bit irrelevant to that bug though, but thanks a lot for the tip, will surely help someone else

**Comment by mostafaberg** [comment]
I've released a new fix that will handle states where peripheral is already connected, can you try that and see if you can still reproduce the bug ? if not i'll need to take another look, let me know
Fix version is `2.1.6`

**Comment by marcussfu** [comment]
update to 2.1.6, still reproduce the bug that stucks on "Connecting".

**Comment by mostafaberg** [comment]
Ok this doesn't reproduce here for some reason, can you supply an example project that reproduces the issue and i'll make sure to get that done right away

**Comment by mostafaberg** [comment]
@marcussfu can confirm i was testing with a completely different mindset on what the issue is, i can confirm this is not fully resolved yet and will try to get the fix out as soon as we can !

**Comment by philips77** [comment]
Hi, we have released version 3.0.1 today. Do you still experience this issue, or can it be closed?

**Comment by sewerynplazuk** [comment]
I'll check it and let you know.

UPDATE:
@philips77, @mostafaberg - in my case the issue has gone. Everything works.
